server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -135,6 +163,64 @@ class LivyServer extends Logging {
     sys.props("livy.server.serverUrl") = _serverUrl.get
   }
 
+  def runKinit(): Boolean = {
+    val keytab = livyConf.get(LAUNCH_KERBEROS_KEYTAB)
+    val principal = SecurityUtil.getServerPrincipal(livyConf.get(LAUNCH_KERBEROS_PRINCIPAL),
+      server.host)
+    require(principal != null,
+      s"Kerberos requires ${LAUNCH_KERBEROS_PRINCIPAL.key} to be provided.")
+    require(keytab != null,
+      s"Kerberos requires ${LAUNCH_KERBEROS_KEYTAB.key} to be provided.")
+    val cacheFileName = kinitCacheDir + "/krb5cc_0"
+    val commands = Seq("kinit", "-kt", keytab, "-c", cacheFileName, principal)
+    val proc = new ProcessBuilder(commands: _*).inheritIO().start()
+    proc.waitFor() match {
+      case 0 =>
+        debug("Ran kinit command successfully.")
+        livyConf.set(RSCConf.Entry.KINIT_CACHE_FILE, cacheFileName)
+        kinitFailCount = 0
+        UserGroupInformation.getCurrentUser.reloginFromTicketCache()
+        true
+      case _ =>
+        warn("Fail to run kinit command," +
+          StringUtils.join(IOUtils.readLines(proc.getErrorStream), "\n"))
+        kinitFailCount += 1
+        false
+    }
+  }
+
+  def startKinitDaemon(): Unit = {
+    executor.scheduleWithFixedDelay(
+      new Runnable() {
+        def run(): Unit = {
+          if (runKinit()) {
+            // schedule another kinit run with a fixed delay.
+            executor.scheduleWithFixedDelay(
```

Comment:
Why are you scheduling a periodic task from within a periodic task? Won't that cause a whole bunch of tasks to execute as time passes?